lib/settingsManager.ts

```diff
@@ -0,0 +1,68 @@
+import { DeepPartial } from "utility-types";
+import { ProjectParser } from "./projectParser";
+import { ISettings, defaultSettings } from "./settingsGenerated";
+export { ISettings };
+
+
+export function normalizeSettings(settings: ISettings) {
+  const newSettings = JSON.parse(JSON.stringify(settings)) as ISettings;
+  for (const [key, value] of Object.entries(newSettings.style)) {
+    if (typeof (newSettings.style as Record<string, unknown>)[key] === 'string') {
+      (newSettings.style as Record<string, unknown>)[key] = (value as string).trim();
+    }
+  }
+  return newSettings;
+}
+
+export function overwriteSettings(settings: ISettings, overwrite: DeepPartial<ISettings>) {
+  const newSettings = JSON.parse(JSON.stringify(settings)) as ISettings;
+  recursiveObjectAssign(newSettings, overwrite);
+  return normalizeSettings(newSettings);
+}
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function recursiveObjectAssign<T extends Record<string, any>>(target: T, source: DeepPartial<T>) {
+  Object.keys(source).forEach(key => {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const s_val = source[key];
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const t_val = target[key];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/strict-boolean-expressions
+    (target as any)[key] = t_val && s_val && typeof t_val === 'object' && typeof s_val === 'object'
+      ? recursiveObjectAssign(t_val, s_val)
+      : s_val;
+  });
+  return target;
+}
+
+export const currentCapabilities = {
+  configuration: false,
+  workspaceFolder: false
+};
+// Cache the settings of all open documents
+export const documentSettings = new Map<string, ISettings>();
```

Comment:
I am wondering if it is wise to make this a singleton.
For running it is probably fine, but for testing it is probably not good.
In the test we generally do not assume to have a global ProjectParser instance. But if you call getDocumentSettings with different projectParser instances I would assume you get funny results. Maybe add the documentSettings to the Projectparser itself?
or just add this function also to projectParser. This just feels weird.

Reply:
Moved to projectParser